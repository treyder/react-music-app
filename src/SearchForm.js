import React from 'react';
import { connect } from "react-redux";
import Button from 'material-ui/Button';

import SearchOptions from './SearchOptions';
import SearchResult from './SearchResult';

import { fetchMovies, searchTextChange, searchByChange } from './actions/searchActions';

export class SearchForm extends React.Component {

    constructor(props) {
        super(props);
        
        this.baseServiceUrl = "http://react-cdp-api.herokuapp.com";
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        this.props.dispatch(fetchMovies(this.baseServiceUrl, this.props.searchBy, this.props.searchText));
    }

    render() {
        const { error, loading, movies, searchText } = this.props;

        if (error) {
            return <div>Error! {error.message}</div>;
        }

        if (loading) {
            return <div>Loading...</div>;
        }

        return (
            <form onSubmit={this.handleSubmit}>
                <input type="text" id="searchText" onChange={event => this.props.dispatch(searchTextChange(event.target.value))} className="Search-input"/>
                <div className="Search-options-table">
                    <div className="Search-options-row">
                        <div className="Search-options-cell">
                            <SearchOptions actionOnChange={event => this.props.dispatch(searchByChange(event.searchBy))} />
                        </div>
                        <div className="Search-options-cell">
                            <Button id="submitButton" variant="raised" color="primary" type="submit">SEARCH</Button>
                        </div>
                    </div>
                </div>
                <SearchResult moviesList={this.props.movies} />
            </form>
        );
    }
};

function mapStateToProps(state) {
    return {
        movies: state.searchReducer.movies,
        loading: state.searchReducer.loading,
        error: state.searchReducer.error,
        searchText: state.searchReducer.searchText,
        searchBy: state.searchReducer.searchBy
    }
};

export default connect(mapStateToProps)(SearchForm);