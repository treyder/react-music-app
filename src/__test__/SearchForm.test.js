import "isomorphic-fetch";
import React from "react";
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { SearchForm } from "../SearchForm";
import rootReducer from "../reducers";

const middlewares = [thunk];
const store = configureMockStore(middlewares);

describe("SearchForm", () => {

    it("renders initial form state", () => {
        const tree = renderer.create(
            <SearchForm store={store}/>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("display error on failure", () => {
        const errorMessage = {message: '404'};
        let searchForm = shallow(<SearchForm error={errorMessage} />);
        expect(searchForm.html()).toEqual("<div>Error! 404</div>");
    });

    it("display loading", () => {
        let searchForm = shallow(<SearchForm loading={true} />);
        expect(searchForm.html()).toEqual("<div>Loading...</div>");
    });

    it("executes search", async () => {
        const dispatchMock = jest.fn();
        let searchForm = shallow(<SearchForm dispatch={dispatchMock} store={store} searchText="lord" searchBy="title" />);
        await searchForm.instance().handleSubmit({ preventDefault: jest.fn() });
        expect(dispatchMock).toBeCalled();
    });

    it("changes search text", () => {
        const dispatchMock = jest.fn();
        let searchForm = shallow(<SearchForm dispatch={dispatchMock} store={store} searchText="" searchBy="title" />);
        searchForm.find('#searchText'). simulate('change', {target: {name: "searchText", value: "lord"}});
        expect(dispatchMock).toBeCalled();
    });

    it("changes search option for form", () => {
        const dispatchMock = jest.fn();
        let searchForm = mount(<SearchForm dispatch={dispatchMock} store={store} searchText="" searchBy="title" />);
        searchForm.find('Button#titleButton').simulate('click');
        expect(dispatchMock).toBeCalled();
    });

});