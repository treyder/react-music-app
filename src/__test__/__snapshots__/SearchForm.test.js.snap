// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders initial form state 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    className="Search-input"
    id="searchText"
    onChange={[Function]}
    type="text"
  />
  <div
    className="Search-options-table"
  >
    <div
      className="Search-options-row"
    >
      <div
        className="Search-options-cell"
      >
        <div
          className="Search-options-table"
        >
          <div
            className="Search-options-row"
          >
            <div
              className="Search-options-cell"
            >
              SEARCH BY
            </div>
            <button
              className="MuiButtonBase-root-16 MuiButton-root-1 MuiButton-raised-6 MuiButton-raisedPrimary-7 MuiButton-sizeSmall-13"
              disabled={false}
              id="titleButton"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-2"
              >
                TITLE
              </span>
              <span
                className="MuiTouchRipple-root-18"
              />
            </button>
            <button
              className="MuiButtonBase-root-16 MuiButton-root-1 MuiButton-raised-6 MuiButton-raisedSecondary-8 MuiButton-sizeSmall-13"
              disabled={false}
              id="genreButton"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-2"
              >
                GENRE
              </span>
              <span
                className="MuiTouchRipple-root-18"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="Search-options-cell"
      >
        <button
          className="MuiButtonBase-root-16 MuiButton-root-1 MuiButton-raised-6 MuiButton-raisedPrimary-7"
          disabled={false}
          id="submitButton"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-2"
          >
            SEARCH
          </span>
          <span
            className="MuiTouchRipple-root-18"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="Search-result"
  >
    Start searching ...
  </div>
</form>
`;
